File: Trees_Graphs/go/Tree.go
package types


import (
  "fmt"
)


type TreeNode struct {
  Value int
  LeftChild *TreeNode
  RightChild *TreeNode
}


func getChild(array []int) (*TreeNode) {
  length := len(array)
  node := &TreeNode{Value: array[length / 2]}

  if length == 1 {
    return node
  }

  if length == 2 {
    node.LeftChild = getChild(array[:length / 2])
  } else {
    node.LeftChild, node.RightChild = getChild(array[:length / 2]), getChild(array[length / 2 + 1:])
  }

  return node
}


func CreateTree(array []int) *TreeNode {
  length := len(array)
  root := &TreeNode{Value: array[length / 2]}

  root.LeftChild, root.RightChild = getChild(array[:length / 2]), getChild(array[length / 2  + 1:])

  return root
}


type LinkedListTreeNode struct {
  Node *TreeNode
  Next *LinkedListTreeNode
}


type LinkedListOfTreeNodes struct{
  Head *LinkedListTreeNode
  Tail *LinkedListTreeNode
}


func (list *LinkedListOfTreeNodes) Append(node *TreeNode) {
  newNode := LinkedListTreeNode{Node: node}
  if list.Head == nil {
    list.Head = &newNode
    list.Tail = &newNode
  } else {
    list.Tail.Next = &newNode
    list.Tail = list.Tail.Next
  }
}


func PrintLinkedListOfTreeNodes(list LinkedListOfTreeNodes) {
  node := list.Head
  for node != nil {
    fmt.Print(node.Node.Value, "->")
    node = node.Next
  }
  fmt.Print("nil\n")
}


func writeDepth(node *TreeNode, dNum int, list *[]*LinkedListOfTreeNodes) {
  var depthList LinkedListOfTreeNodes
  if dNum == len(*list) {
    depthList.Append(node)
    *list = append(*list, &depthList)
  } else {
    depthList = *(*list)[dNum]
    depthList.Append(node)
    (*list)[dNum] = &depthList
  }

  if node.LeftChild != nil {
    writeDepth(node.LeftChild, dNum + 1, list)
  }

  if node.RightChild != nil {
    writeDepth(node.RightChild, dNum + 1, list)
  }
}


func PrintTreeNodes(root *TreeNode) {
  var depths []*LinkedListOfTreeNodes
  var depthList LinkedListOfTreeNodes

  depthList.Append(root)
  depths = append(depths, &depthList)

  writeDepth(root.LeftChild, 1, &depths)
  writeDepth(root.RightChild, 1, &depths)

  for i := 0; i < len(depths); i++ {
    PrintLinkedListOfTreeNodes(*depths[i])
  }
}

File: Trees_Graphs/python/Tree.py
from LinkedList import *


class TreeNode(object):
    def __init__(self, value, parent=None):
        self.value = value
        self.left_child = None
        self.right_child = None
        self.parent = parent

    def __str__(self):
        return str(self.value)

    def __repr__(self):
        return str(self.value)


class Tree():
    def __init__(self, array):
        length = len(array)
        root = TreeNode(array[int(length / 2)])

        root.left_child, root.right_child = self.get_child(array[:int(length / 2)], root), self.get_child(array[int(length / 2)  + 1:], root)
        self.root = root

    def get_child(self, array, parent):
        length = len(array)
        node = TreeNode(array[int(length / 2)], parent)

        if length == 1:
            return node

        if length == 2:
            node.left_child = self.get_child(array[:int(length / 2)], node)
        else:
            node.left_child, node.right_child = self.get_child(array[:int(length / 2)], node), self.get_child(array[int(length / 2)  + 1:], node)

        return node

    def write_depth(self, node, depth_num, depths):
        depth_list = LinkedList()

        if depth_num == len(depths):
            depth_list.append(node)
            depths.append(depth_list)
        else:
            depth_list = depths[depth_num]
            depth_list.append(node)
            depths[depth_num] = depth_list

        if node.left_child:
            self.write_depth(node.left_child, depth_num + 1, depths)

        if node.right_child:
            self.write_depth(node.right_child, depth_num + 1, depths)

    def print_tree(self):
        depths = []
        depth_list = LinkedList()

        depth_list.append(self.root)

        depths.append(depth_list)

        self.write_depth(self.root.left_child, 1, depths)
        self.write_depth(self.root.right_child, 1, depths)

        for depth in depths:
            print(depth)


class LinkedListTreeNode(object):
    def __init__(self):
        self.node = None
        self.next = None


class LinkedListOfTreeNodes(object):
    def __init__(self):
        self.head = None
        self.tail = None
